ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/SetCodeTransactionProcessor.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright contributors to Hyperledger Besu.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.mainnet;
+
+import org.hyperledger.besu.crypto.SignatureAlgorithm;
+import org.hyperledger.besu.crypto.SignatureAlgorithmFactory;
+import org.hyperledger.besu.datatypes.Address;
+import org.hyperledger.besu.datatypes.Hash;
+import org.hyperledger.besu.datatypes.SetCodeAuthorization;
+import org.hyperledger.besu.ethereum.core.Transaction;
+import org.hyperledger.besu.ethereum.core.encoding.SetCodeTransactionEncoder;
+import org.hyperledger.besu.ethereum.rlp.BytesValueRLPOutput;
+import org.hyperledger.besu.evm.account.MutableAccount;
+import org.hyperledger.besu.evm.worldstate.WorldUpdater;
+
+import java.math.BigInteger;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+
+import com.google.common.base.Suppliers;
+import org.apache.tuweni.bytes.Bytes;
+
+public class SetCodeTransactionProcessor {
+  private static final Bytes MAGIC = Bytes.of(0x05);
+
+  private static final Supplier<SignatureAlgorithm> SIGNATURE_ALGORITHM =
+      Suppliers.memoize(SignatureAlgorithmFactory::getInstance);
+
+  private final BigInteger chainId;
+
+  public SetCodeTransactionProcessor(final BigInteger chainId) {
+    this.chainId = chainId;
+  }
+
+  public Set<Address> addContractToAuthority(
+      final WorldUpdater worldUpdater, final Transaction transaction) {
+    final Set<Address> authorityList = new HashSet<>();
+
+    transaction
+        .setCodeTransactionPayloads()
+        .get()
+        .forEach(
+            payload -> {
+              recoverAuthority(payload)
+                  .ifPresent(
+                      authorityAddress -> {
+                        if (!chainId.equals(BigInteger.ZERO)
+                            && !payload.chainId().equals(BigInteger.ZERO)
+                            && !chainId.equals(payload.chainId())) {
+                          return;
+                        }
+
+                        final MutableAccount account = worldUpdater.getAccount(authorityAddress);
+
+                        if (payload.nonces().size() == 1
+                            && !payload.nonces().getFirst().equals(account.getNonce())) {
+                          return;
+                        }
+
+                        worldUpdater.addCodeToEOA(
```

Comment:
should we check the nonce before we add code to eoa?  i.e. why set code if the nonce is already invalid